### PR TITLE
Make MetricsLogger as a component

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #945

MetricsLogger should be a component as its role is similar to CheckpointManager, which provides some functions and has its own states. More importantly, users may want to customize the metrics. Make it a component and can be customized through TrainSpec.

Change the name of `MetricsLogger` to `MetricsProcessor` as it not only log but also process metrics.

### CODE REVIEW

File: torchtitan/train.py [hunk 1]
```diff
@@ -350,6 +336,10 @@ def main(job_config: JobConfig):
                 train_state.step == 1
                 or train_state.step % job_config.metrics.log_freq == 0
             ):
+                # NOTE: Loss info after the last log step before checkpoint saving will
+                # not be ploted. This can be avoided by setting checkpoint.interval to
+                # be a multiple of metrics.log_freq
+
```
Comment (tianyu-l) [inline torchtitan/train.py]: I think this message is no longer relevant, after we change loss logging to be the loss at a log step, instead of avg of the last `log_freq` interval, so we can remove.

File: torchtitan/train.py [hunk 2]
```diff
@@ -132,9 +125,14 @@ def main(job_config: JobConfig):
     model_converters = build_model_converters(job_config, parallel_dims)
     model_converters.convert(model)
 
+    # metrics logging
+    build_metrics_logger_fn = build_metrics_logger if train_spec.build_metrics_logger_fn is None else traiin_spec.build_metrics_logger_fn
```
Comment (tianyu-l) [inline torchtitan/train.py]:
```suggestion
    build_metrics_logger_fn = build_metrics_logger if train_spec.build_metrics_logger_fn is None else train_spec.build_metrics_logger_fn
```

File: torchtitan/train.py [hunk 3]
```diff
@@ -285,8 +269,10 @@ def main(job_config: JobConfig):
             data_load_start = time.perf_counter()
             batch = next(data_iterator)
             input_ids, labels = batch
-            ntokens_since_last_log += labels.numel()
-            data_loading_times.append(time.perf_counter() - data_load_start)
+            metrics_logger.ntokens_since_last_log += labels.numel()
+            metrics_logger.data_loading_times.append(
+                time.perf_counter() - data_load_start
+            )
```
Comment (tianyu-l) [inline torchtitan/train.py]: I can see you are doing a hard job to make it as isolated as possible.

But uh, this means customized metric logger also must have these fields ...

File: torchtitan/train.py [hunk 4]
```diff
@@ -363,61 +353,7 @@ def main(job_config: JobConfig):
                 else:
                     global_avg_loss = global_max_loss = loss.item()
 
-                # update train state
-                train_state.log_steps.append(train_state.step)
-                train_state.global_avg_losses.append(global_avg_loss)
-                train_state.global_max_losses.append(global_max_loss)
```
Comment (tianyu-l) [inline torchtitan/train.py:377]: shall we just remove these fields in `train_state`, and remove the logging from checkpoint option at all?

**Review by tianyu-l — COMMENTED** [review]
overall looks good to me!
I wish we can be less prescriptive on what fields a customized metric logger need to have, but I don't have a better way...

File: torchtitan/train.py (hunk 4 above)
Comment (fegin) [reply torchtitan/train.py:377]: ye, good point, I split this PR from the trainer PR. That should already be removed. Let me update this PR.

File: torchtitan/train.py (hunk 2 above)
Comment (fegin) [reply torchtitan/train.py:129]: Thanks. I should add UT to catch the else case.

File: torchtitan/train.py (hunk 3 above)
Comment (fegin) [reply torchtitan/train.py]: I think logically speaking, these fields belong to `MetricsLogger` because `data_loading_times` is technically a metric. And Trainer just happens to be the one that calculate the metric and sets the data to `MetricsLogger`.

**Comment by fegin** [comment]
@tianyu-l Logically speaking, those fields belong to `MetricsLogger`. I was wondering whether I should remove `Logger` postfix and just call it `Metrics`. With `Logger` prefix, it feels that the fields may belong to some other components like Trainer and `MetricsLogger` is just used to log the metrics. On the other hand, if we call it `Metrics`, this looks more sense as `Metrics` should hold all the metrics and have action to log them.

File: torchtitan/train.py (hunk 2 above)
Comment (fegin) [reply torchtitan/train.py]: Thanks, I think we should have a UT for the training loop. Let me do this after I make Trainer as a class.

File: torchtitan/train.py [hunk 5]
```diff
@@ -237,18 +243,6 @@ def main(job_config: JobConfig):
         return
 
     checkpoint.load(step=job_config.checkpoint.load_step)
-    metric_logger = build_metric_logger(job_config, parallel_dims)
-
-    # plot losses loaded from checkpoint (if any) to TensorBoard
-    # NOTE: Loss info after the last log step before checkpoint saving will not be ploted.
-    #       This can be avoided by setting checkpoint.interval to be a multiple of metrics.log_freq
-    if train_state.step > 0 and not job_config.metrics.disable_logging_from_checkpoint:
```
Comment (fduwjj) [inline torchtitan/train.py:253]: if you decide to remove it, then can we also remove `disable_logging_from_checkpoint`?

File: torchtitan/components/metrics.py [hunk 6]
```diff
@@ -244,3 +246,129 @@ def build_metric_logger(
...
+        self.ntokens_since_last_log = 0
+        self.data_loading_times.clear()
+        self.time_last_log = time.perf_counter()
+        self.device_memory_monitor.reset_peak_stats()
+
+    def close(self):
+        self.logger.close()
+
+
+def build_metrics_logger(
```
Comment (fduwjj) [inline torchtitan/components/metrics.py]: So to override or change it, I now need to define a different `class MetricsLogger` and a different function build_metrics_logger so that I can do whatever metrics logging I want?

Comment (fegin) [reply torchtitan/components/metrics.py]: ye, you can inherit `MetricsLogger`, but if you hate inheritance, then you can completely rewrite, :)

**Review by fduwjj — APPROVED** [review]
unblock
